submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them.
+
+The proposal has its own set of Pros and Cons.
+
+## Pros
+
+Since we will be annotating only select K8s objects with globalIp, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIp,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIp at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary CRDs which create unique
+globalIps at the Namespace/Pod level.
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec               EgressIPSpec `json:"spec"`
+}
+
+type EgressIPSpec struct {
+    // Selects the pods to which the GlobalnetEgressIP will be applied.
+    // If multiple pods match the selector the same egressIP will be used for all those pods.
+    // +optional
+    PodSelector PodSelectorSpec `json:"podSelector,omitempty"`
+
+    // Clients can request a specific globalIP to be allocated. If the requested
+    // globalIP is part of globalCIDR of the Cluster and is not allocated to other
+    // K8s objects, then it will be allocated, otherwise creation of the CRD will fail.
+    // Valid values are empty string (""), or a valid IP address from the globalCIDR of
+    // the Cluster. This field cannot be changed through updates.
+    GlobalIP    string          `json:"private_ip"`
+}
+
+type PodSelectorSpec struct {
+    // MatchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+    // map is equivalent to an element of matchExpressions, whose key field is "key", the
+    // operator is "In", and the values array contains only "value". The requirements are ANDed.
+    MatchLabels map[string]string `json:"matchLabels"`
+
+    // Associated priority to be used for the selected pods.
+    // Its value should be between 0 (lowest) and 1000 (highest) inclusive.
+    // When this value is omitted, the default value of 0 is used.
+    // +optional
+    Priority    int32 `json:"priority,omitempty"`
```

Comment:
Because of the LabelSelectors change, let's move this priority up to the EgressIPSpec

Reply:
As you know, IPTables do not have priority fields like openflow rules. So care has to be taken to program rules in the right order to ensure priority is retained. Also using separate Chains (like a chain name for ClusterLevelGlobalIP, NameSpaceLevelGlobalIP and a chain for PodLevelGlobalIPs) will make debugging/investigating issues a bit easier. 

So when users explicitly require IPs for Pods, they will carry higher priority over namespace. So, priority is a field that goes along with PodSelector.

Reply:
But won't that order of iptables will be the same regardless of where we put the priority field (here or in the GlobalnetEgressIP) ? may be I'm missing something

Reply:
From an implementation POV, my plan is to create Separate Chains for each level.
For example: SUBMARINER-GN-PODS, SUBMARINER-GN-NAMESPACES and SUBMARINER-GN-CLUSTER

So, for egress traffic destined to remoteClusters we will have to use a combination of marking and ipSets to redirect traffic as follows.

```
+--------------------+
|                    |
|  Egress Traffic    +----+
|                    |    |
+--------------------+    |
                          |
                          v
         +----------------+---+
         |                    |
         | SUBMARINER-GN-PODS +-------+
         |                    |       |
         +--------------------+       |
                                      |
                                      v
                        +-------------+---------+
                        |                       |
                        | SUBMARINER-GN-NAMESPACE
                        |                       |
                        +---------------------+-+
                                              |
                                              v
                                +-------------+-------+
                                |                     |
                                | SUBMARINER-GN-CLUSTER
                                |                     |
                                +---------------------+
```

The order of iptables (based on the priority) will matter mostly in the `SUBMARINER-GN-PODS` chain. If we put the priority in the `Spec`, it can apply to even namespace based GlobalnetEgressIP object (which does not have any podSelectors)

i.e., 
```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
     namespace: myproject
   spec:
     priority: 100            <-- This will be confusing to users.
```

Happy to explain over a call if required.

Reply:
Ahh, I see what you mean. But you could still create an empty podselector, with just priority. I see what you mean though.

Reply:
I say let's drop the priority altogether

Reply:
Why no `NamespaceSelector` For the use case to allocate GlobalIP to namespace, won't it be better to create a GlobalnetEgressIP object with NamespaceSelector? I realize that scope namespace with empty podSelector can achieve that, but thinking from an ownership perspective. Who decides which namespaces to get their own globalIP, user owning the namespace or admin owning the cluster? Thinking about RBAC implications of the two approaches.

Reply:
Done